Change some aspects of the string-to-date conversion with format to make it more similar to the SQL standard #2388

TreeHunter9 opened the pull request:
When converting a formatted character string to a datetime, the default values for the date parts are taken from `CURRENT_TIMESTAMP`.
Change **(A/P)M** to **(A/P).M.** for a 12-hour period.
Fix **RM** pattern.
Add more unit test for conversion function.

TreeHunter9 committed:
Use current TimeStamp for data in stringToDate conversion if it's not specify

Also fix RM pattern and change (A/P)M to (A/P).M.

TreeHunter9 committed:
Add more tests

TreeHunter9 committed:
Add TimeStamp validation

Also move duplicated code to functions.

TreeHunter9 committed:
Add more unit tests for "YY" and "YYY" patterns

TreeHunter9 commented:
I'm using some sort of "hack" in CvtTest.cpp by redefining `TimeStamp::getCurrentTimeStamp()` for tests, and can someone more experienced with C++ say is this even correct?

asfernandes commented:
> I'm using some sort of "hack" in CvtTest.cpp by redefining `TimeStamp::getCurrentTimeStamp()` for tests, and can someone more experienced with C++ say is this even correct?

IMO this is problematic.

It probably only work due to the way the project is split in common static library, that is implementation detail.

It may also be problematic if this method becomes inline.

Mock in C++ usually is done with static (templates) or dynamic (virtual methods) polymorphism:
- static may make the code difficult, usually everything must be done in header
- dynamic may be slow

I think none of them is appropriate for `TimeStamp::getCurrentTimeStamp()`, but you may instead make the formatter code asks for it trough an indirection (for example, an std::function) that you may replace.

TreeHunter9 committed:
Use Callback for getting current date

It's better because we can mock Callback for unit tests.

TreeHunter9 commented:
> I think none of them is appropriate for `TimeStamp::getCurrentTimeStamp()`, but you may instead make the formatter code asks for it trough an indirection (for example, an std::function) that you may replace.

Yeah, it would work, but I looked for an alternative of `TimeStamp::getCurrentTimeStamp()` and found `getLocalDate()` from `Callbacks`. I think it's perfect fit for mocking, and I already kinda mocked it in **CvtTestUtils.h**.

TreeHunter9 committed:
Fix exception and README description

TreeHunter9 committed:
Add ability to print blr_cast_format

TreeHunter9 committed:
Put a comment about new BLR in the right place

dyemanov commented:
Are there any changes in this branch that's still pending? Or were all suggestions addressed?

Commit:
Add information about behavior of string to datetime conversion

TreeHunter9 commented:
I reread that issue and @mrotteveel said that `HH` and `HH12` implemented differently from the standard (A.M. and P.M. should be in pattern, not in formatting string, if I understand it correctly).

mrotteveel commented:
> I re-read that issue and @mrotteveel said that `HH` and `HH12` implemented differently from the standard (A.M. and P.M. should be in pattern, not in formatting string, if I understand it correctly).

Yes, the current implementation doesn't implement `<datetime template am/pm>`, and `<datetime template 12-hour>` should not render `A.M.` or `P.M.` after the time, that is solely the responsibility of `<datetime template am/pm>`

dyemanov commented:
So let's have AM/PM fixed in this branch and only then merge it.

Commit:
Rework old patterns and add new ones

Add A.M, P.M., RR and RRRR patterns.
Rework YY, YYY, HH and HH12 patterns due to new patterns.
Add restriction from SQL standard to format.
Fix incorrect error message for mismatched pattern.
Fix bug with 0 hours in HH12.

Commit:
Add more unit tests

Commit:
Update doc for cast format

Commit:
Allow specification of log_level for BOOST_TESTS in make

TreeHunter9 commented:
I also added <datetime template rounded year> pattern, but I'm not sure if it matches the standard. I read it multiple times, and still don't get what the standard wants from me. I use Oracle description of <datetime template rounded year> and added it to docs.
@mrotteveel it seems you have some experience of reading and understanding ISO standard, can you please check if Oracle description of <datetime template rounded year> is similar to standard one?

dyemanov commented:
@asfernandes, @mrotteveel any more comments / suggestions for this PR?

mrotteveel commented:
I'll try to look at it tomorrow, but it will more likely be in the form of building from the branch and trying some things out than looking at the code.

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -2037,9 +2313,50 @@ static void string_to_format_datetime_pattern_matcher(std::string_view pattern,
+	// If CT contains <datetime template second of day>, then CT shall not contain any of the following:
+	// <datetime template 12-hour>, <datetime template 24-hour>, <datetime template minute>, <datetime
+	// template second of minute>, or <datetime template am/pm>.
+	if (static_cast<bool>(formatFlags & FormatPatterns::SSSSS))
```
Are these `static_cast<bool>` necessary?

TreeHunter9 replied:
Yes. I recently read an interesting [discussion](https://chat.stackoverflow.com/rooms/9345/discussion-between-nicol-bolas-and-potatoswatter) about `enum class`, and maybe using it as bitfield wasn't a good idea.
It's interesting to hear what other people think about `enum class` as bitfield, maybe I should change it to simple `enum`?

aafemt replied:
For bitfields I would use constexpressions in namespace. Using enum class for it is strange.

TreeHunter9 replied:
> For bitfields I would use constexpressions in namespace.

Yea, it's used a lot in our code base, but I don't know, it doesn't seem very informative when you looking at `int` variable in my opinion.
P.S. Sorry for misleading, **bitfields** is something different, I mean **bitflags**, but i think you get the point.

TreeHunter9 replied:
Oh, that's an interesting approach, first I thought about unnamed namespace, but now I get it.

aafemt replied:
Also I would call the namespace "Format". In this case `Format::Pattern` and `Format::SSSSS` would be closer to so-called "literate programming". Code readability and self-documenting in one.

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -136,6 +136,76 @@ const SINT64 INT64_LIMIT = ((((SINT64) 1) << 62) / 5);
 // AP,2012: Looks like there is no need making len zero, but I keep old define for a reference.
 // {if (len) {memcpy(to, from, len); from += len; to += len; len = 0;} }
 
+#define DEFINE_ENUM_BITWISE_OPERATORS(enumType) \
```
I would move this defines to some generic header as it could be useful for others `enum class` code and `cvt.cpp` already has many things.

TreeHunter9 replied:
Done

Commit:
Change enum class to enum in namespace

Commit:
Switch from plain enum to constexpr values

mrotteveel commented:
> @asfernandes, @mrotteveel any more comments / suggestions for this PR?

Unfortunately, I didn't have time or energy yet to look at it in detail. Hopefully later this week, but otherwise maybe it should just be merged and we can fix/correct things later.
